osu.Game/Beatmaps/BeatmapManager_UpdateQueue.cs

```diff
@@ -0,0 +1,180 @@
+                {
+                    try
+                    {
+                        using (var db = new SqliteConnection(storage.GetDatabaseConnectionString("online")))
+                        {
+                            var found = db.QueryFirstOrDefault<CachedOnlineBeatmapLookup>(
+                                "SELECT * FROM osu_beatmaps WHERE checksum = @MD5Hash OR beatmap_id = @OnlineBeatmapID OR filename = @Path", beatmap);
+
+                            if (found != null)
+                            {
+                                var status = (BeatmapSetOnlineStatus)found.approved;
+
+                                beatmap.Status = status;
+                                beatmap.BeatmapSet.Status = status;
+                                beatmap.BeatmapSet.OnlineBeatmapSetID = found.beatmapset_id;
+                                beatmap.OnlineBeatmapID = found.beatmap_id;
+
+                                LogForModel(set, $"Cached local retrieval for {beatmap}.");
+                                return;
+                            }
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        LogForModel(set, $"Cached local retrieval for {beatmap} failed with {ex}.");
+                    }
+                }
+
+                if (api?.State != APIState.Online)
+                    return;
+
+                var req = new GetBeatmapRequest(beatmap);
+
+                req.Failure += fail;
+
+                try
+                {
+                    // intentionally blocking to limit web request concurrency
+                    api.Perform(req);
+
+                    var res = req.Result;
+
+                    if (res != null)
+                    {
+                        beatmap.Status = res.Status;
+                        beatmap.BeatmapSet.Status = res.BeatmapSet.Status;
+                        beatmap.BeatmapSet.OnlineBeatmapSetID = res.OnlineBeatmapSetID;
+                        beatmap.OnlineBeatmapID = res.OnlineBeatmapID;
+
+                        LogForModel(set, $"Online retrieval mapped {beatmap} to {res.OnlineBeatmapSetID} / {res.OnlineBeatmapID}.");
+                    }
+                }
+                catch (Exception e)
+                {
+                    fail(e);
+                }
+
+                void fail(Exception e)
+                {
+                    beatmap.OnlineBeatmapID = null;
+                    LogForModel(set, $"Online retrieval failed for {beatmap} ({e.Message})");
+                }
+            }
+
+            private void prepareLocalCache()
+            {
+                string cacheFilePath = storage.GetFullPath(cache_database_name);
+                string compressedCacheFilePath = $"{cacheFilePath}.bz2";
+
+                cacheDownloadRequest = new FileWebRequest(compressedCacheFilePath, $"https://assets.ppy.sh/client-resources/{cache_database_name}.bz2");
```

Comment:
So as a general question, is this a one-and-done sort of thing for now, or is this going to be periodically updated? Because if this file will get updated, it'd be nice to automatically refetch it every now and again to improve cache hit rate for new beatmaps.

Reply:
As per the PR description, that will be added at a future point. The main goal is to reduce queries on initial import, and even after months or years it will still suffice from that perspective. Still not 100% on how this should be distributed. nuget package in consumer projects seems like the best path forward in my mind (having it bundled with updates and install) as it will benefit from binary diffing for updates and not require its own download/update/patch logic.